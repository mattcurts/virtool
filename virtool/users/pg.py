from __future__ import annotations

from datetime import datetime
from typing import List

from sqlalchemy import (
    Boolean,
    Index,
)
from sqlalchemy import Table, Column, ForeignKey
from sqlalchemy.ext.associationproxy import AssociationProxy
from sqlalchemy.orm import Mapped, mapped_column, relationship

from virtool.groups.pg import SQLGroup
from virtool.pg.base import Base


class UserGroup(Base):
    __tablename__ = "user_group"

    user_id: Mapped[int] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), primary_key=True
    )
    group_id: Mapped[int] = mapped_column(
        ForeignKey("groups.id", ondelete="CASCADE"), primary_key=True
    )

    is_primary: Mapped[bool] = mapped_column(default=False)

    group: Mapped["SQLGroup"] = relationship(lazy="joined")
    user: Mapped["SQLUser"] = relationship(back_populates="user_group_associations")

    Index(
        "primary_group_unique",
        is_primary,
        user_id,
        unique=True,
        postgresql_where=(is_primary == True),
    ),


class SQLUser(Base):
    __tablename__ = "users"

    id: Mapped[int] = mapped_column(primary_key=True)
    legacy_id: Mapped[str | None] = mapped_column(unique=True)
    active: Mapped[bool] = mapped_column(default=True)
    administrator: Mapped[bool] = mapped_column(default=False)
    b2c_display_name: Mapped[str] = mapped_column(default="")
    b2c_given_name: Mapped[str] = mapped_column(default="")
    b2c_family_name: Mapped[str] = mapped_column(default="")
    b2c_oid: Mapped[str] = mapped_column(default="")
    force_reset: Mapped[bool] = mapped_column(default=False)
    handle: Mapped[str]
    invalidate_sessions: Mapped[bool] = mapped_column(default=False)
    last_password_change: Mapped[datetime]
    password: Mapped[bytes | None]

    user_group_associations: Mapped[List[UserGroup]] = relationship(
        back_populates="user", cascade="all, delete-orphan", lazy="joined"
    )

    groups: AssociationProxy[List[SQLGroup]] = AssociationProxy(
        "user_group_associations",
        "group",
        creator=lambda group: UserGroup(group=group),
    )

    primary_group_association: Mapped[UserGroup] = relationship(
        back_populates="user",
        lazy="joined",
        primaryjoin="and_(user_group.c.user_id == SQLUser.id, user_group.c.is_primary == True)",
        viewonly=True,
    )

    primary_group: AssociationProxy[SQLGroup] = AssociationProxy(
        "primary_group_association",
        "group",
    )

    def to_dict(self):
        base_dict = super().to_dict()
        base_dict["groups"] = self.groups
        base_dict["primary_group"] = self.primary_group
        return base_dict

    def __repr__(self):
        params = ", ".join(
            f"{column}='{type(value).__name__ if column == 'last_password_change' else value}'"
            for column, value in self.to_dict().items()
            if column not in ["password"]
        )

        return f"<{self.__class__.__name__}({params})>"
