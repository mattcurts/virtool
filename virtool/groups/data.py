import asyncio
import math
from typing import List

from sqlalchemy import delete, select, func
from sqlalchemy.exc import IntegrityError
from sqlalchemy.ext.asyncio import AsyncEngine, AsyncSession
from virtool_core.models.group import GroupMinimal, Group, GroupSearchResult
from virtool_core.models.user import UserNested


from virtool.authorization.client import AuthorizationClient
from virtool.data.errors import ResourceNotFoundError, ResourceConflictError
from virtool.data.events import emits, Operation, emit
from virtool.data.topg import both_transactions
from virtool.data.transforms import apply_transforms
from virtool.groups.mongo import (
    update_member_users_and_api_keys,
)
from virtool.groups.oas import UpdateGroupRequest
from virtool.groups.pg import SQLGroup
from virtool.mongo.core import Mongo
from virtool.users.utils import generate_base_permissions
from virtool.utils import base_processor

class GroupsData:
    name = "groups"

    def __init__(
        self, authorization_client: AuthorizationClient, mongo: Mongo, pg: AsyncEngine
    ):
        self._authorization_client = authorization_client
        self._mongo = mongo
        self._pg = pg

    async def list(self) -> List[GroupMinimal]:
        """
        List all user groups.

        :return: a list of all user groups

        """
        async with AsyncSession(self._pg) as session:
            result = await session.execute(select(SQLGroup))
            return [GroupMinimal(**group.to_dict()) for group in result.scalars()]

    async def find(
        self, page: int, per_page: int, paginate=False, term: str | None = None
    ) -> List[GroupMinimal]:
        """
        Finds all user groups matching the term

        :return: a list of all user groups

        """
        if paginate:
            return await self._find_beta(page, per_page, term)

        return await self.list()

    async def _find_beta(
        self, page: int, per_page: int, term: str | None = None
    ) -> GroupSearchResult:

        total_query = (
            select(func.count(SQLGroup.id).label("total"))
            .subquery()
        )

        found_query = (
            select(func.count(SQLGroup.id).label("found"))
            .where(SQLGroup.name.ilike(f"%{term}%"))
            .subquery()
        )

        skip = 0

        if page > 1:
            skip = (page - 1) * per_page

        async with AsyncSession(self._pg) as session:
            query = (
                select(SQLGroup)
                .offset(skip)
                .limit(per_page)
            )

            total_count_results, found_count_results, results = await asyncio.gather(
                session.execute(select(total_query)),
                session.execute(select(found_query)),
                session.execute(query),
            )

            total_count = total_count_results.scalar()
            found_count = found_count_results.scalar()

            groups = [row.to_dict() for row in results.unique().scalars()]

        return GroupSearchResult(
            items=groups,
            found_count=found_count,
            total_count=total_count,
            page=page,
            page_count=int(math.ceil(found_count / per_page)),
            per_page=per_page,
        )

    async def get(self, group_id: int) -> Group:
        """
        Get a single group by its ID.

        :param group_id: the group's ID
        :return: the group
        """

        async with AsyncSession(self._pg) as session:
            group: SQLGroup | None = await session.get(SQLGroup, group_id)

            if group:
                users = [
                    UserNested(**base_processor(user))
                    async for user in self._mongo.users.find({"groups": group_id})
                ]

                return Group(
                    **{
                        **group.to_dict(),
                        "users": users,
                    }
                )

        raise ResourceNotFoundError

    @emits(Operation.CREATE)
    async def create(self, name: str) -> Group:
        """
        Create new group with the given name.

        :param name: the id for the new group
        :return: the group
        :raises ResourceConflictError: if a group with the given name already exists
        """
        try:
            async with AsyncSession(self._pg) as session:
                group = SQLGroup(
                    legacy_id=None,
                    name=name,
                    permissions=generate_base_permissions(),
                )

                session.add(group)

                await session.flush()

                group_id = group.id

                await session.commit()
        except IntegrityError:
            raise ResourceConflictError("Group already exists")

        return await self.get(group_id)

    @emits(Operation.UPDATE)
    async def update(self, group_id: int, data: UpdateGroupRequest) -> Group:
        """
        Update the name or permissions for a group.

        :param group_id: the id of the group
        :param data: updates to the current group permissions or name
        :return: the updated group
        :raises ResourceNotFoundError: if the group does not exist

        """

        async with both_transactions(self._mongo, self._pg) as (
            mongo_session,
            pg_session,
        ):
            group = await pg_session.get(SQLGroup, group_id)

            if not group:
                raise ResourceNotFoundError

            data = data.dict(exclude_unset=True)

            db_update = {}

            if "name" in data:
                group.name = data["name"]

            if "permissions" in data:
                group.permissions = {**group.permissions, **data["permissions"]}

            if db_update:
                group.update(db_update)
                await update_member_users_and_api_keys(
                    self._mongo, mongo_session, pg_session, group.id
                )

        return await self.get(group_id)

    async def delete(self, group_id: int):
        """
        Delete a group by its id.

        Deletes the group and updates all member user permissions if they are affected
        by deletion of the group.

        :param group_id: the id of the group to delete
        :raises ResourceNotFoundError: if the group is not found

        """
        group = await self.get(group_id)

        async with both_transactions(self._mongo, self._pg) as (
            mongo_session,
            pg_session,
        ):
            result = await pg_session.execute(
                delete(SQLGroup).where(SQLGroup.id == group_id)
            )

            if not result.rowcount:
                raise ResourceNotFoundError

            await update_member_users_and_api_keys(
                self._mongo, mongo_session, pg_session, group_id
            )

        emit(group, "groups", "delete", Operation.DELETE)
