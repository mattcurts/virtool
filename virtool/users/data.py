import asyncio

from pymongo.errors import DuplicateKeyError
from sqlalchemy import select, update, delete, insert
from sqlalchemy.ext.asyncio import AsyncEngine
from virtool_core.models.roles import AdministratorRole
from virtool_core.models.user import User, UserSearchResult

import virtool.users.utils
import virtool.utils
from virtool.administrators.db import update_legacy_administrator
from virtool.api.utils import compose_regex_query, paginate_aggregate
from virtool.authorization.client import AuthorizationClient
from virtool.authorization.relationships import AdministratorRoleAssignment
from virtool.data.domain import DataLayerDomain
from virtool.data.errors import ResourceConflictError, ResourceNotFoundError
from virtool.data.events import emits, Operation
from virtool.data.topg import compose_legacy_id_expression, both_transactions
from virtool.data.transforms import apply_transforms
from virtool.errors import DatabaseError
from virtool.groups.pg import merge_group_permissions, SQLGroup
from virtool.groups.transforms import AttachPrimaryGroupTransform, AttachGroupsTransform
from virtool.mongo.core import Mongo
from virtool.mongo.utils import id_exists
from virtool.users.db import (
    B2CUserAttributes,
    compose_groups_update,
)
from virtool.users.mongo import (
    create_user,
    update_keys,
    compose_primary_group_update,
    generate_handle,
)
from virtool.users.oas import UpdateUserRequest
from virtool.users.pg import SQLUser, user_group_associations
from virtool.users.transforms import AttachPermissionsTransform
from virtool.utils import base_processor

PROJECTION = [
    "_id",
    "handle",
    "administrator",
    "force_reset",
    "groups",
    "last_password_change",
    "permissions",
    "primary_group",
    "administrator_role",
    "active",
    "b2c",
    "b2c_display_name",
    "b2c_family_name",
    "b2c_given_name",
    "b2c_oid",
]


class UsersData(DataLayerDomain):
    name = "users"

    def __init__(
        self, authorization_client: AuthorizationClient, mongo: Mongo, pg: AsyncEngine
    ):
        self._authorization_client = authorization_client
        self._mongo = mongo
        self._pg = pg

    async def find(
        self,
        page: int,
        per_page: int,
        administrator: bool | None = None,
        term: str | None = None,
    ) -> UserSearchResult:
        """
        Find users.

        Optionally filter by administrator status or search term.

        :param page: the page number
        :param per_page: the number of items per page
        :param administrator: whether to filter by administrator status
        :param term: a search term to filter by user handle
        """

        administrator_roles = dict(
            await self._authorization_client.list_administrators()
        )

        administrator_query = {}

        if administrator is not None:
            operator = "$in" if administrator else "$nin"
            administrator_query = {"_id": {operator: list(administrator_roles.keys())}}

        term_query = compose_regex_query(term, ["handle"]) if term else {}

        client_query = {**administrator_query, **term_query}

        result = await paginate_aggregate(
            self._mongo.users,
            page,
            per_page,
            client_query,
            sort="handle",
            projection={field: True for field in PROJECTION},
        )

        result["items"] = await apply_transforms(
            [base_processor(item) for item in result["items"]],
            [
                AttachPermissionsTransform(self._pg),
                AttachPrimaryGroupTransform(self._pg),
                AttachGroupsTransform(self._pg),
            ],
        )

        result["items"] = [
            User(**user, administrator_role=administrator_roles.get(user["id"]))
            for user in result["items"]
        ]

        return UserSearchResult(**result)

    async def get(self, user_id: str) -> User:
        """
        Get a user by their ``user_id``.

        :param user_id: the user's ID
        :return: the user
        """
        user, (user_id, role) = await asyncio.gather(
            self._mongo.users.find_one(
                {"_id": user_id},
            ),
            self._authorization_client.get_administrator(user_id),
        )

        if not user:
            raise ResourceNotFoundError("User does not exist")

        return User(
            **(
                await apply_transforms(
                    base_processor(user),
                    [
                        AttachPermissionsTransform(self._pg),
                        AttachPrimaryGroupTransform(self._pg),
                        AttachGroupsTransform(self._pg),
                    ],
                )
            ),
            administrator_role=role,
        )

    @emits(Operation.CREATE)
    async def create(
        self,
        handle: str,
        password: str | None,
        force_reset: bool = False,
        b2c_user_attributes: B2CUserAttributes | None = None,
    ) -> User:
        """
        Create a new user.

        If Azure AD B2C information is given, add it to user document.

        :param handle: the requested handle for the user
        :param password: a password
        :param force_reset: force the user to reset password on next login
        :param  b2c_user_attributes: Azure b2c user attributes used to describe a user
        :return: the user document
        """
        async with both_transactions(self._mongo, self._pg) as (mongo, pg):
            now = virtool.utils.timestamp()

            document = await create_user(
                self._mongo,
                handle,
                password,
                force_reset,
                b2c_user_attributes=b2c_user_attributes,
                session=mongo,
            )

            pg.add(
                SQLUser(
                    legacy_id=document["_id"],
                    handle=handle,
                    password=document["password"] if password else None,
                    force_reset=force_reset,
                    last_password_change=now,
                    b2c_display_name=b2c_user_attributes.display_name
                    if b2c_user_attributes
                    else "",
                    b2c_given_name=b2c_user_attributes.given_name
                    if b2c_user_attributes
                    else "",
                    b2c_family_name=b2c_user_attributes.family_name
                    if b2c_user_attributes
                    else "",
                    b2c_oid=b2c_user_attributes.oid if b2c_user_attributes else "",
                )
            )

        return await self.get(document["_id"])

    async def delete(self, legacy_id: str):
        async with both_transactions(self._mongo, self._pg) as (mongo, pg):
            await pg.execute(delete(SQLUser).where(legacy_id=legacy_id))
            await self._mongo.users.deleteOne(_id=legacy_id, session=mongo)

    async def delete_all(self):
        async with both_transactions(self._mongo, self._pg) as (mongo, pg):
            await pg.execute(delete(SQLUser))
            await self._mongo.users.delete_many({}, session=mongo)

    async def create_first(self, handle: str, password: str) -> User:
        """
        Create the first instance user.

        :param handle: the user handle
        :param password: the password
        :return:
        """
        if await self.check_users_exist():
            raise ResourceConflictError("Virtool already has at least one user")

        if handle == "virtool":
            raise ResourceConflictError("Reserved user name: virtool")

        async with both_transactions(self._mongo, self._pg) as (mongo, pg):
            now = virtool.utils.timestamp()

            document = await create_user(
                self._mongo, handle, password, False, session=mongo
            )
            await self._mongo.users.update_one(
                {"_id": document["_id"]},
                {"$set": {"administrator": True}},
                session=mongo,
            )
            user = SQLUser(
                legacy_id=document["_id"],
                handle=handle,
                password=document["password"],
                force_reset=False,
                last_password_change=now,
            )
            pg.add(user)

        await self.set_administrator_role(document["_id"], AdministratorRole.FULL)

        return await self.get(document["_id"])

    async def find_or_create_b2c_user(
        self, b2c_user_attributes: B2CUserAttributes
    ) -> User:
        """
        Search for existing user using an OID.

        If not found, create new user with the OID and user attributes. Auto-generate a
        handle.

        :param b2c_user_attributes: User attributes collected from ID token claims
        :return: the found or created user
        """
        if document := await self._mongo.users.find_one(
            {"b2c_oid": b2c_user_attributes.oid}, ["_id"]
        ):
            return await self.get(document["_id"])

        handle = await generate_handle(
            self._mongo.users,
            b2c_user_attributes.given_name,
            b2c_user_attributes.family_name,
        )

        try:
            user = await self.create(
                handle, None, False, b2c_user_attributes=b2c_user_attributes
            )
        except DuplicateKeyError:
            return await self.find_or_create_b2c_user(b2c_user_attributes)

        return user

    async def set_administrator_role(
        self, user_id: str, role: AdministratorRole
    ) -> User:
        """
        Set a user's administrator role.

        Sets the user's legacy administrator flag to ``True`` if the ``FULL`` user role
        is set. Otherwise, sets the flag to ``False``.

        :param user_id: the id of the user to set the role of
        :param role: the administrator role
        :return: the administrator
        """

        if not await id_exists(self._mongo.users, user_id):
            raise ResourceNotFoundError("User does not exist")

        async with both_transactions(self._mongo, self._pg) as (mongo, pg):
            await update_legacy_administrator(self._mongo, user_id, role, mongo)
            await pg.execute(
                update(SQLUser)
                .where(SQLUser.legacy_id == user_id)
                .values(administrator=role == AdministratorRole.FULL)
            )

            if role is None:
                # Clear the user's administrator role.
                admin_tuple = await self._authorization_client.get_administrator(
                    user_id
                )

                if admin_tuple[1] is not None:
                    await self._authorization_client.remove(
                        AdministratorRoleAssignment(
                            user_id, AdministratorRole(admin_tuple[1])
                        )
                    )
            else:
                await self._authorization_client.add(
                    AdministratorRoleAssignment(user_id, AdministratorRole(role))
                )

        user = await self.get(user_id)

        return user

    # TODO: Update this one to use both_transactions
    @emits(Operation.UPDATE)
    async def update(self, user_id: str, data: UpdateUserRequest):
        """
        Update a user.

        Sessions and API keys are updated as well.

        :param user_id: the ID of the user to update
        :param data: the update data object
        :return: the updated user
        """
        async with both_transactions(self._mongo, self._pg) as (mongo, pg):
            document = await self._mongo.users.find_one(
                {"_id": user_id}, ["administrator", "groups"], session=mongo
            )

            user = (
                await pg.execute(
                    select(SQLUser).where(SQLUser.legacy_id == user_id).limit(1)
                )
            ).scalar()

            if document is None or user is None:
                raise ResourceNotFoundError("User does not exist")

            data = data.dict(exclude_unset=True)

            updates = {}

            if "administrator" in data:
                updates["administrator"] = data["administrator"]
                user.administrator = data["administrator"]

                role_assignment = AdministratorRoleAssignment(
                    user_id, AdministratorRole.FULL
                )

                if data["administrator"]:
                    await self._authorization_client.add(role_assignment)
                else:
                    await self._authorization_client.remove(role_assignment)

            if "force_reset" in data:
                updates.update(
                    {"force_reset": data["force_reset"], "invalidate_sessions": True}
                )
                user.force_reset = data["force_reset"]
                user.invalidate_sessions = True

            if "password" in data:
                hashedpass = virtool.users.utils.hash_password(data["password"])
                timestamp = virtool.utils.timestamp()
                updates.update(
                    {
                        "password": hashedpass,
                        "last_password_change": timestamp,
                        "invalidate_sessions": True,
                    }
                )
                user.password = hashedpass
                user.last_password_change = timestamp
                user.invalidate_sessions = True

            if "groups" in data:
                try:
                    updates.update(
                        await compose_groups_update(self._pg, data["groups"])
                    )
                    for _id in data["groups"]:
                        await pg.execute(
                            insert(user_group_associations).values(
                                user_id=user.id, group_id=_id
                            )
                        )

                except DatabaseError as err:
                    raise ResourceConflictError(str(err))

            if "primary_group" in data:
                try:
                    primary_group = await compose_primary_group_update(
                        self._mongo,
                        self._pg,
                        data.get("groups", []),
                        data["primary_group"],
                        user_id,
                    )

                except DatabaseError as err:
                    raise ResourceConflictError(str(err))
                updates.update(primary_group)
                user.primary_group = primary_group.values()

            if "active" in data:
                updates.update({"active": data["active"], "invalidate_sessions": True})
                user.active = data["active"]
                user.invalidate_sessions = True

            if updates:
                document = await self._mongo.users.find_one_and_update(
                    {"_id": user_id}, {"$set": updates}, session=mongo
                )

                if document["groups"]:
                    result = await pg.execute(
                        select(SQLGroup).where(
                            compose_legacy_id_expression(SQLGroup, document["groups"])
                        )
                    )

                    groups = [group.to_dict() for group in result.scalars().all()]
                else:
                    groups = []

                await update_keys(
                    self._mongo,
                    user_id,
                    document["administrator"],
                    document["groups"],
                    merge_group_permissions(groups),
                    session=mongo,
                )
        return await self.get(user_id)

    async def check_users_exist(self) -> bool:
        """
        Checks that users exist.

        :returns: True if users exist otherwise False
        """
        return await self._mongo.users.count_documents({}, limit=1) > 0
