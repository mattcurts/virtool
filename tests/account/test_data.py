import asyncio
import datetime

import pytest
from aiohttp.test_utils import make_mocked_coro
from syrupy.filters import props
from syrupy.matchers import path_type
from virtool_core.models.enums import Permission
from virtool_core.models.roles import AdministratorRole

from virtool.account.oas import CreateKeysRequest, UpdateAccountRequest
from virtool.groups.oas import PermissionsUpdate
from virtool.pg.utils import get_row_by_id
from virtool.users.pg import SQLUser

_last_password_change_matcher = path_type(
    {"last_password_change": (datetime.datetime,)}
)


@pytest.mark.parametrize(
    "administrator_role", [AdministratorRole.FULL, None], ids=["full", "none"]
)
@pytest.mark.parametrize(
    "has_permission", [True, False], ids=["has permission", "missing permission"]
)
async def test_create_api_key(
    administrator_role,
    has_permission,
    mocker,
    mongo,
    snapshot,
    static_time,
    data_layer,
    fake2,
):
    """
    Test that an API key is created correctly with varying key owner administrator status and
    permissions.

    """
    mocker.patch("virtool.account.db.get_alternate_id", make_mocked_coro("foo_0"))

    mocker.patch("virtool.utils.generate_key", return_value=("bar", "baz"))

    group_1 = await fake2.groups.create()
    group_2 = await fake2.groups.create(
        PermissionsUpdate(
            **{
                Permission.create_sample: True,
                Permission.modify_subtraction: has_permission,
            }
        )
    )

    user = await fake2.users.create(
        groups=[group_1, group_2], administrator_role=administrator_role
    )

    _, api_key = await data_layer.account.create_key(
        CreateKeysRequest(
            name="Foo",
            permissions=PermissionsUpdate(create_sample=True, modify_subtraction=True),
        ),
        user.id,
    )

    assert api_key == snapshot(name="dl")
    assert await mongo.keys.find_one() == snapshot(name="mongo")


@pytest.mark.parametrize(
    "update",
    [
        UpdateAccountRequest(old_password="hello_world_1", password="hello_world_2"),
        UpdateAccountRequest(email="hello@world.com"),
        UpdateAccountRequest(
            old_password="hello_world_1",
            password="hello_world_2",
            email="hello@world.com",
        ),
    ],
    ids=["password", "email", "password and email"],
)
async def test_update(data_layer, mongo, pg, fake2, snapshot, update):
    user = await fake2.users.create(password="hello_world_1")

    await data_layer.account.update(
        user.id,
        update,
    )

    (row, document) = await asyncio.gather(
        get_row_by_id(pg, SQLUser, 1), mongo.users.find_one({"_id": user.id})
    )

    assert row == snapshot(
        name="pg", matcher=_last_password_change_matcher, exclude=props("password")
    )
    assert document == snapshot(
        name="mongo", matcher=_last_password_change_matcher, exclude=props("password")
    )
    assert row.password == document["password"]
