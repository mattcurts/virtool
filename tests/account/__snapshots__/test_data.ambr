# serializer version: 1
# name: test_create_api_key[has permission-full][dl]
  dict({
    'administrator': True,
    'created_at': datetime.datetime(2015, 10, 6, 20, 0),
    'groups': list([
      dict({
        'id': 2,
        'legacy_id': None,
        'name': 'hydrogeologists',
      }),
      dict({
        'id': 1,
        'legacy_id': None,
        'name': 'musicians',
      }),
    ]),
    'id': 'foo_0',
    'name': 'Foo',
    'permissions': dict({
      'cancel_job': False,
      'create_ref': False,
      'create_sample': True,
      'modify_hmm': False,
      'modify_subtraction': True,
      'remove_file': False,
      'remove_job': False,
      'upload_file': False,
    }),
  })
# ---
# name: test_create_api_key[has permission-full][mongo]
  dict({
    '_id': 'baz',
    'administrator': True,
    'created_at': datetime.datetime(2015, 10, 6, 20, 0),
    'groups': list([
      2,
      1,
    ]),
    'id': 'foo_0',
    'name': 'Foo',
    'permissions': dict({
      'cancel_job': False,
      'create_ref': False,
      'create_sample': True,
      'modify_hmm': False,
      'modify_subtraction': True,
      'remove_file': False,
      'remove_job': False,
      'upload_file': False,
    }),
    'user': dict({
      'id': 'bf1b993c',
    }),
  })
# ---
# name: test_create_api_key[has permission-none][dl]
  dict({
    'administrator': False,
    'created_at': datetime.datetime(2015, 10, 6, 20, 0),
    'groups': list([
      dict({
        'id': 2,
        'legacy_id': None,
        'name': 'hydrogeologists',
      }),
      dict({
        'id': 1,
        'legacy_id': None,
        'name': 'musicians',
      }),
    ]),
    'id': 'foo_0',
    'name': 'Foo',
    'permissions': dict({
      'cancel_job': False,
      'create_ref': False,
      'create_sample': True,
      'modify_hmm': False,
      'modify_subtraction': True,
      'remove_file': False,
      'remove_job': False,
      'upload_file': False,
    }),
  })
# ---
# name: test_create_api_key[has permission-none][mongo]
  dict({
    '_id': 'baz',
    'administrator': False,
    'created_at': datetime.datetime(2015, 10, 6, 20, 0),
    'groups': list([
      2,
      1,
    ]),
    'id': 'foo_0',
    'name': 'Foo',
    'permissions': dict({
      'cancel_job': False,
      'create_ref': False,
      'create_sample': True,
      'modify_hmm': False,
      'modify_subtraction': True,
      'remove_file': False,
      'remove_job': False,
      'upload_file': False,
    }),
    'user': dict({
      'id': 'bf1b993c',
    }),
  })
# ---
# name: test_create_api_key[missing permission-full][dl]
  dict({
    'administrator': True,
    'created_at': datetime.datetime(2015, 10, 6, 20, 0),
    'groups': list([
      dict({
        'id': 2,
        'legacy_id': None,
        'name': 'hydrogeologists',
      }),
      dict({
        'id': 1,
        'legacy_id': None,
        'name': 'musicians',
      }),
    ]),
    'id': 'foo_0',
    'name': 'Foo',
    'permissions': dict({
      'cancel_job': False,
      'create_ref': False,
      'create_sample': True,
      'modify_hmm': False,
      'modify_subtraction': True,
      'remove_file': False,
      'remove_job': False,
      'upload_file': False,
    }),
  })
# ---
# name: test_create_api_key[missing permission-full][mongo]
  dict({
    '_id': 'baz',
    'administrator': True,
    'created_at': datetime.datetime(2015, 10, 6, 20, 0),
    'groups': list([
      2,
      1,
    ]),
    'id': 'foo_0',
    'name': 'Foo',
    'permissions': dict({
      'cancel_job': False,
      'create_ref': False,
      'create_sample': True,
      'modify_hmm': False,
      'modify_subtraction': True,
      'remove_file': False,
      'remove_job': False,
      'upload_file': False,
    }),
    'user': dict({
      'id': 'bf1b993c',
    }),
  })
# ---
# name: test_create_api_key[missing permission-none][dl]
  dict({
    'administrator': False,
    'created_at': datetime.datetime(2015, 10, 6, 20, 0),
    'groups': list([
      dict({
        'id': 2,
        'legacy_id': None,
        'name': 'hydrogeologists',
      }),
      dict({
        'id': 1,
        'legacy_id': None,
        'name': 'musicians',
      }),
    ]),
    'id': 'foo_0',
    'name': 'Foo',
    'permissions': dict({
      'cancel_job': False,
      'create_ref': False,
      'create_sample': True,
      'modify_hmm': False,
      'modify_subtraction': False,
      'remove_file': False,
      'remove_job': False,
      'upload_file': False,
    }),
  })
# ---
# name: test_create_api_key[missing permission-none][mongo]
  dict({
    '_id': 'baz',
    'administrator': False,
    'created_at': datetime.datetime(2015, 10, 6, 20, 0),
    'groups': list([
      2,
      1,
    ]),
    'id': 'foo_0',
    'name': 'Foo',
    'permissions': dict({
      'cancel_job': False,
      'create_ref': False,
      'create_sample': True,
      'modify_hmm': False,
      'modify_subtraction': False,
      'remove_file': False,
      'remove_job': False,
      'upload_file': False,
    }),
    'user': dict({
      'id': 'bf1b993c',
    }),
  })
# ---
# name: test_update[email][mongo]
  dict({
    '_id': 'bf1b993c',
    'active': True,
    'administrator': False,
    'email': 'hello@world.com',
    'force_reset': False,
    'groups': list([
    ]),
    'handle': 'leeashley',
    'invalidate_sessions': False,
    'last_password_change': datetime,
    'primary_group': None,
    'settings': dict({
      'quick_analyze_workflow': 'pathoscope_bowtie',
      'show_ids': True,
      'show_versions': True,
      'skip_quick_analyze_dialog': True,
    }),
  })
# ---
# name: test_update[email][pg]
  <SQLUser(id='1', legacy_id='bf1b993c', active='True', administrator='False', b2c_display_name='', b2c_given_name='', b2c_family_name='', b2c_oid='', force_reset='False', handle='leeashley', invalidate_sessions='False', last_password_change='datetime', groups='[]', primary_group='None')>
# ---
# name: test_update[password and email][mongo]
  dict({
    '_id': 'bf1b993c',
    'active': True,
    'administrator': False,
    'email': 'hello@world.com',
    'force_reset': False,
    'groups': list([
    ]),
    'handle': 'leeashley',
    'invalidate_sessions': False,
    'last_password_change': datetime,
    'primary_group': None,
    'settings': dict({
      'quick_analyze_workflow': 'pathoscope_bowtie',
      'show_ids': True,
      'show_versions': True,
      'skip_quick_analyze_dialog': True,
    }),
  })
# ---
# name: test_update[password and email][pg]
  <SQLUser(id='1', legacy_id='bf1b993c', active='True', administrator='False', b2c_display_name='', b2c_given_name='', b2c_family_name='', b2c_oid='', force_reset='False', handle='leeashley', invalidate_sessions='False', last_password_change='datetime', groups='[]', primary_group='None')>
# ---
# name: test_update[password][mongo]
  dict({
    '_id': 'bf1b993c',
    'active': True,
    'administrator': False,
    'force_reset': False,
    'groups': list([
    ]),
    'handle': 'leeashley',
    'invalidate_sessions': False,
    'last_password_change': datetime,
    'primary_group': None,
    'settings': dict({
      'quick_analyze_workflow': 'pathoscope_bowtie',
      'show_ids': True,
      'show_versions': True,
      'skip_quick_analyze_dialog': True,
    }),
  })
# ---
# name: test_update[password][pg]
  <SQLUser(id='1', legacy_id='bf1b993c', active='True', administrator='False', b2c_display_name='', b2c_given_name='', b2c_family_name='', b2c_oid='', force_reset='False', handle='leeashley', invalidate_sessions='False', last_password_change='datetime', groups='[]', primary_group='None')>
# ---
